Implement globalnet for headless service without selector

mkimuram opened the pull request:
This PR adds support for globalnet for headless service without selector.

Fixes: #1537 

Dependency:
- RBAC:
  - [x] submariner-io/submariner-operator#1582
  - [x] submariner-io/submariner-charts#169
  - [x] #1582

mkimuram commented:
@sridhargaddam @vthapar  @aswinsuryan 

This is still a prototype that only supports a single IP of the endpoints (and many quick & dirty hacks remaining). However, it seems to work (See below for details). Let's make this prototype as a start point of the further discussion.

<details>

1. Deploy this branch

```bash
make e2e using=external-net,globalnet
```

2. Update RBAC to allow handling core/v1 endpoints for both cluster1 and cluster2 ([this](https://github.com/submariner-io/submariner-operator/blob/devel/bundle/manifests/submariner-globalnet_rbac.authorization.k8s.io_v1_clusterrole.yaml#L13) should be modified, later.)

(Just kubectl edit the ClusterRole submariner-globalnet to add `endpoints` in `resources` of `apiGroups == ""`)

```bash
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1
kubectl get ClusterRole submariner-globalnet -o yaml > /tmp/rbac-before.yaml 
kubectl edit ClusterRole submariner-globalnet
kubectl get ClusterRole submariner-globalnet -o yaml > /tmp/rbac-after.yaml 
diff -urN /tmp/rbac-before.yaml /tmp/rbac-after.yaml 
--- /tmp/rbac-before.yaml       2021-09-16 23:34:30.754644486 +0000
+++ /tmp/rbac-after.yaml        2021-09-16 23:35:13.890396448 +0000
@@ -3,7 +3,7 @@
 metadata:
   creationTimestamp: "2021-09-16T23:31:06Z"
   name: submariner-globalnet
-  resourceVersion: "859"
+  resourceVersion: "1952"
   uid: 14770711-cda4-47a7-85d3-8cb750a394be
 rules:
 - apiGroups:
@@ -13,6 +13,7 @@
   - services
   - namespaces
   - nodes
+  - endpoints
   verbs:
   - get
   - list
```

(Do the same to cluster2)

```bash
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
kubectl edit ClusterRole submariner-globalnet
```

3. Test headless service without selector

- Create headless service without selector and endpoints, and export the service
```bash
docker inspect "ext-app" -f '{{(index .NetworkSettings.Networks "'pseudo-ext'").IPAddress}}'
ip=$(docker inspect "ext-app" -f '{{(index .NetworkSettings.Networks "'pseudo-ext'").IPAddress}}')
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1 
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: test-vm
spec:
  clusterIP: None
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
EOF

cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-vm
subsets:
  - addresses:
      - ip: ${ip}
    ports:
      - port: 80
EOF

subctl export service -n default test-vm
```

- Check the globalnet log
```bash
kubectl logs -n submariner-operator -l app=submariner-globalnet

I0917 16:42:44.860635       1 service_export_controller.go:147] Processing ServiceExport "default/test-vm"
I0917 16:42:44.961770       1 ingress_endpoints_controller.go:93] Created Endpoints controller for (default/test-vm) with selector "metadata.name=test-vm"
I0917 16:42:44.962547       1 ingress_endpoints_controller.go:125] "create" ingress Endpoints default/test-vm for service test-vm
I0917 16:42:44.972396       1 global_ingressip_controller.go:135] Processing created &v1.GlobalIngressIP{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"ep-test-vm", GenerateName:"", Namespace:"default", SelfLink:"", UID:"a163ca0d-125c-4f80-a0ad-c471c0cf6536", ResourceVersion:"2307", Generation:1, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63767493764, loc:(*time.Location)(0x2026ea0)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"submariner.io/serviceRef":"test-vm"}, Annotations:map[string]string{"submariner.io/headless-svc-endpoints-ip":"172.117.0.4"}, OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:"", ManagedFields:[]v1.ManagedFieldsEntry{v1.ManagedFieldsEntry{Manager:"submariner-globalnet", Operation:"Update", APIVersion:"submariner.io/v1", Time:(*v1.Time)(0xc0006a0db0), FieldsType:"FieldsV1", FieldsV1:(*v1.FieldsV1)(0xc0006a0dc8)}}}, Spec:v1.GlobalIngressIPSpec{Target:"HeadlessServiceEndpoints", ServiceRef:(*v1.LocalObjectReference)(0xc0005c7070), PodRef:(*v1.LocalObjectReference)(0xc0005c7080)}, Status:v1.GlobalIngressIPStatus{Conditions:[]v1.Condition(nil), AllocatedIP:""}}
I0917 16:42:44.972455       1 global_ingressip_controller.go:158] Allocating global IP for "default/ep-test-vm"
I0917 16:42:44.972476       1 iface.go:145] Installing iptables rule for Headless SVC Pod -d 242.254.1.253 -j DNAT --to 172.117.0.4
I0917 16:42:44.976890       1 iface.go:227] Installing iptable egress rules for HDLS SVC Pod "default/ep-test-vm": -p all -s 172.117.0.4 -m mark --mark 0xC0000/0xC0000 -j SNAT --to 242.254.1.253
I0917 16:42:44.981070       1 base_controllers.go:194] Updated: &v1.GlobalIngressIPStatus{Conditions:[]v1.Condition{v1.Condition{Type:"Allocated", Status:"True", ObservedGeneration:0, LastTransitionTime:v1.Time{Time:time.Time{wall:0xc0494f413a79ca9a, ext:311044842452, loc:(*time.Location)(0x2026ea0)}}, Reason:"Success", Message:"Allocated global IP"}}, AllocatedIP:"242.254.1.253"}
```

- Confirm that globalingress IP is assigned to the service
```bash
kubectl get globalingressip ep-test-vm -o jsonpath='{.status.allocatedIP}{"\n"}'
242.254.1.253
```

- Test Access from cluster1 and cluster2, and check the log

```bash
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1 
kubectl -n default run tmp-shell --rm -i --tty --image quay.io/submariner/nettest -- bash
curl 242.254.1.253
```

```bash
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2 
kubectl -n default run tmp-shell --rm -i --tty --image quay.io/submariner/nettest -- bash
curl 242.254.1.253
```

```bash
docker logs ext-app
172.117.0.3 - - [17/Sep/2021 16:44:17] "GET / HTTP/1.1" 200 -
242.254.2.6 - - [17/Sep/2021 16:44:37] "GET / HTTP/1.1" 200 -
```

</details>

mkimuram commented:
Also, manually confirmed that reverse access also works as expected.

<details>

- Deploy StatefulSet in cluster2

```bash
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: nginx-ss
  labels:
    app.kubernetes.io/instance: nginx-ss
    app.kubernetes.io/name: nginx-ss
spec:
  ports:
  - port: 80
    name: web
  clusterIP: None
  selector:
    app.kubernetes.io/instance: nginx-ss
    app.kubernetes.io/name: nginx-ss
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  serviceName: "nginx-ss"
  replicas: 2
  selector:
    matchLabels:
        app.kubernetes.io/instance: nginx-ss
        app.kubernetes.io/name: nginx-ss
  template:
    metadata:
      labels:
        app.kubernetes.io/instance: nginx-ss
        app.kubernetes.io/name: nginx-ss
    spec:
      containers:
      - name: nginx-ss
        image: k8s.gcr.io/nginx-slim:0.8
        ports:
        - containerPort: 80
          name: web
EOF
subctl export service -n default nginx-ss
```

- Check globalingressip for each pod managed by the StatefulSet
```bash
kubectl get globalingressip pod-web-0 -o jsonpath='{.status.allocatedIP}{"\n"}'
242.254.2.252
```

```bash
kubectl get globalingressip pod-web-1 -o jsonpath='{.status.allocatedIP}{"\n"}'
242.254.2.251
```

- Test access from external app to each pod

```bash
docker exec -it ext-app bash
curl 242.254.2.252
curl 242.254.2.251
```

- Check the access log for each pod (source IPs match ext-app's globalIngressIP)
```bash
kubectl logs web-0
242.254.1.253 - - [17/Sep/2021:19:26:29 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
kubectl logs web-1
242.254.1.253 - - [17/Sep/2021:19:26:33 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
```

- Test access from each pod to external app

```bash
kubectl exec -it web-0 -- bash
curl 242.254.1.253
```

```bash
kubectl exec -it web-1 -- bash
curl 242.254.1.253
```

- Check the access log from each pod (source IPs match each pod's globalIngressIP)

```bash
docker logs ext-app
242.254.2.252 - - [17/Sep/2021 19:29:55] "GET / HTTP/1.1" 200 -
242.254.2.251 - - [17/Sep/2021 19:30:15] "GET / HTTP/1.1" 200 -
```

</details>

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,153 @@
+
+	ingressIP := &submarinerv1.GlobalIngressIP{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("ep-%.59s", endpoints.Name),
```
```suggestion
			Name:      fmt.Sprintf("ep-%.60s", endpoints.Name),
```

I think we can go with 60 here.

mkimuram replied:
Fixed.

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,153 @@
+
+	if op == syncer.Delete {
+		c.ingressIPMap.Remove(ingressIP.Name)
+		klog.Infof("ingress Endpoints %s for service %s deleted", key, c.svcName)
```
```suggestion
		klog.Infof("Ingress Endpoints %s for service %s deleted", key, c.svcName)
```

mkimuram replied:
Fixed.

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,153 @@
+	klog.Infof("%q ingress Endpoints %s for service %s", op, key, c.svcName)
+
+	// TODO: consider handling multiple endpoints IPs
+	if len(endpoints.Subsets) == 0 || len(endpoints.Subsets[0].Addresses) == 0 {
```
As per this implementation, if a user manually creates an endpoint with more than one ipAddress, we will only program rules for the first entry right?

mkimuram replied:
Yes. The current implementation only uses the first entry.

This commit meant to be just a quick prototype. I expect that multiple entries can be added, however I'm not sure how multiple `GlobalIngressIP`s can be handled by using admiral's syncer framework. Can it just return `GlobalIngressIPList`? (and then need to create a controller for it?)

mkimuram replied:
Note that this is still TODO.

sridhargaddam replied:
> Yes. The current implementation only uses the first entry.
> 
> This commit meant to be just a quick prototype. I expect that multiple entries can be added, however I'm not sure how multiple `GlobalIngressIP`s can be handled by using admiral's syncer framework. Can it just return `GlobalIngressIPList`? (and then need to create a controller for it?)

@tpantelis can you please advise?

mkimuram replied:
@sridhargaddam @anfredette @nerdalert 

> As per this implementation, if a user manually creates an endpoint with more than one ipAddress, we will only program rules for the first entry right?

The last commit adds support for multiple IPs in endpoints of headless service without selector. PTAL
(Note that first 2 commits will be merged through #1582 )

Current E2E test doesn't cover this specific case, but I've manually tested by below way and worked well:

<details>

1. Deploy the test environment

```
make deploy using=external-net,globalnet
```

2. Create headless service without selector with one endpoint and test connectivity and source IPs

- In cluster1, create headless service without selector with one endpoint
```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1
 
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: test-vm
spec:
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
  clusterIP: None
EOF
 
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-vm
subsets:
  - addresses:
      - ip: 172.117.0.4
    ports:
      - port: 80
EOF
 
 
kubectl get svc,endpoints,globalingressip
NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   100.0.0.1    <none>        443/TCP   23m
service/test-vm      ClusterIP   None         <none>        80/TCP    27s
 
NAME                   ENDPOINTS         AGE
endpoints/kubernetes   172.18.0.5:6443   23m
endpoints/test-vm      172.117.0.4:80    27s
 
 
subctl export service -n default test-vm
```

- globalingressip is created after the service is exported
```
kubectl get svc,endpoints,globalingressip
NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   100.0.0.1    <none>        443/TCP   24m
service/test-vm      ClusterIP   None         <none>        80/TCP    55s
 
NAME                   ENDPOINTS         AGE
endpoints/kubernetes   172.18.0.5:6443   24m
endpoints/test-vm      172.117.0.4:80    55s
 
NAME                                                   IP
globalingressip.submariner.io/ep-test-vm-172.117.0.4   242.254.1.253
 
kubectl get -o yaml globalingressip.submariner.io/ep-test-vm-172.117.0.4
apiVersion: submariner.io/v1
kind: GlobalIngressIP
metadata:
  annotations:
    submariner.io/headless-svc-endpoints-ip: 172.117.0.4
  creationTimestamp: "2022-03-08T22:48:58Z"
  generation: 2
  labels:
    submariner-io/originatingNamespace: default
    submariner.io/serviceRef: test-vm
  name: ep-test-vm-172.117.0.4
  namespace: default
  resourceVersion: "5297"
  uid: 46ed1f13-e65c-4240-8374-0935313d134b
spec:
  serviceRef:
    name: test-vm
  target: HeadlessServiceEndpoints
status:
  allocatedIP: 242.254.1.253
  conditions:
  - lastTransitionTime: "2022-03-08T22:48:58Z"
    message: Allocated global IP
    reason: Success
    status: "True"
    type: Allocated
``` 
 
- In cluster2, create statefulset, service without selector and export the service.
```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
 
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: nginx-ss
  labels:
    app.kubernetes.io/instance: nginx-ss
    app.kubernetes.io/name: nginx-ss
spec:
  ports:
  - port: 80
    name: web
  clusterIP: None
  selector:
    app.kubernetes.io/instance: nginx-ss
    app.kubernetes.io/name: nginx-ss
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  serviceName: "nginx-ss"
  replicas: 2
  selector:
    matchLabels:
        app.kubernetes.io/instance: nginx-ss
        app.kubernetes.io/name: nginx-ss
  template:
    metadata:
      labels:
        app.kubernetes.io/instance: nginx-ss
        app.kubernetes.io/name: nginx-ss
    spec:
      containers:
      - name: nginx-ss
        image: k8s.gcr.io/nginx-slim:0.8
        ports:
        - containerPort: 80
          name: web
EOF
 
subctl export service -n default nginx-ss
```

- Test access from pod to external
```
kubectl get globalingressip
NAME        IP
pod-web-0   242.254.2.252
pod-web-1   242.254.2.253
 
kubectl exec -it web-0 -- curl -m 10 242.254.1.253
 
docker logs ext-app 2>&1 | tail -n 1
242.254.2.252 - - [08/Mar/2022 22:56:55] "GET / HTTP/1.1" 200 -
 
kubectl exec -it web-1 -- curl -m 10 242.254.1.253
 
docker logs ext-app 2>&1 | tail -n 1
242.254.2.253 - - [08/Mar/2022 22:57:40] "GET / HTTP/1.1" 200 -
```

- Test access from external to pod
```
docker exec -it ext-app curl -m 10 242.254.2.252
 
kubectl logs web-0
242.254.1.253 - - [08/Mar/2022:22:59:11 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
 
 
docker exec -it ext-app curl -m 10 242.254.2.253
kubectl logs web-1
242.254.1.253 - - [08/Mar/2022:23:00:12 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
```

3. Create another test container, add it to endpoints, and test connectivity and source IPs

- Create another test container
```
docker run --cap-add=NET_ADMIN --name ext-app2 --network pseudo-ext \
-d registry.access.redhat.com/ubi7/ubi python -m SimpleHTTPServer 80
 
docker exec -i ext-app2 yum install -y iproute
docker exec -i ext-app2 ip r add 242.254.1.0/24 via 172.117.0.3
docker exec -i ext-app2 ip r add 242.254.2.0/24 via 172.117.0.3
```

- Add it to endpoints (Then new ingressglobalip ep-test-vm-172.117.0.5 is added)
```
docker inspect ext-app2 -f '{{(index .NetworkSettings.Networks "pseudo-ext").IPAddress}}'
172.117.0.5
 
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1
 
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-vm
subsets:
  - addresses:
      - hostname: ext-app
        ip: 172.117.0.4
      - hostname: ext-app2
        ip: 172.117.0.5
    ports:
      - port: 80
EOF
 
kubectl get svc,endpoints,globalingressip
NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   100.0.0.1    <none>        443/TCP   52m
service/test-vm      ClusterIP   None         <none>        80/TCP    28m
 
NAME                   ENDPOINTS                       AGE
endpoints/kubernetes   172.18.0.5:6443                 52m
endpoints/test-vm      172.117.0.4:80,172.117.0.5:80   28m
 
NAME                                                   IP
globalingressip.submariner.io/ep-test-vm-172.117.0.4   242.254.1.253
globalingressip.submariner.io/ep-test-vm-172.117.0.5   242.254.1.252
```
 
- Test access from pod to external
```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
 
kubectl exec -it web-0 -- curl -m 10 242.254.1.252
 
docker logs ext-app2 2>&1 | tail -n 1
242.254.2.252 - - [08/Mar/2022 23:19:05] "GET / HTTP/1.1" 200 -
 
kubectl exec -it web-1 -- curl -m 10 242.254.1.252
 
docker logs ext-app2 2>&1 | tail -n 1
242.254.2.253 - - [08/Mar/2022 23:19:25] "GET / HTTP/1.1" 200 -
```

- Test access from external to pod
```
docker exec -it ext-app2 curl -m 10 242.254.2.252
 
kubectl logs web-0
242.254.1.253 - - [08/Mar/2022:22:59:11 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
242.254.1.252 - - [08/Mar/2022:23:20:31 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
 
docker exec -it ext-app2 curl -m 10 242.254.2.253
kubectl logs web-1
242.254.1.253 - - [08/Mar/2022:23:00:12 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
242.254.1.252 - - [08/Mar/2022:23:20:42 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
```

</details>

sridhargaddam commented on pkg/globalnet/controllers/global_ingressip_controller.go:
```diff
@@ -88,6 +90,15 @@ func NewGlobalIngressIPController(config syncer.ResourceSyncerConfig, pool *ipam
+				key, _ := cache.MetaNamespaceKeyFunc(obj)
+				return controller.iptIface.AddEgressRulesForHeadlessSVCPods(key, target, reservedIPs[0], globalNetIPTableMark)
+			} else if gip.Spec.Target == submarinerv1.HeadlessServiceEndpoints {
+				// TODO: fixme (Maybe only rename and change messages?)
+				err := controller.iptIface.AddIngressRulesForHeadlessSvcPod(reservedIPs[0], target)
+				if err != nil {
+					return err
+				}
+
```
This is same as L88-L94. I agree, we can rename it and match both Target with submarinerv1.HeadlessServicePod || submarinerv1.HeadlessServiceEndpoints

sridhargaddam commented on pkg/globalnet/controllers/global_ingressip_controller.go:
```diff
@@ -207,6 +218,40 @@ func (c *globalIngressIPController) onCreate(ingressIP *submarinerv1.GlobalIngre
+			return true
+		}
+	} else if ingressIP.Spec.Target == submarinerv1.HeadlessServiceEndpoints {
+		endpointsIP := ingressIP.GetAnnotations()[headlessSvcEndpointsIP]
+		if endpointsIP == "" {
+			_ = c.pool.Release(ips...)
+
+			klog.Warningf("%q annotation is missing on %q", headlessSvcEndpointsIP, key)
+
+			return true
+		}
+
+		// TODO: fixme (Just rename and modifying messages?)
+		err = c.iptIface.AddIngressRulesForHeadlessSvcPod(ips[0], endpointsIP)
+		if err != nil {
+			klog.Errorf("Error while programming Service %q ingress rules for Pod: %v", key, err)
+			err = errors.WithMessage(err, "Error programming ingress rules")
+		} else {
+			err = c.iptIface.AddEgressRulesForHeadlessSVCPods(key, endpointsIP, ips[0], globalNetIPTableMark)
+			if err != nil {
+				_ = c.iptIface.RemoveIngressRulesForHeadlessSvcPod(ips[0], endpointsIP)
+				err = errors.WithMessage(err, "Error programming egress rules")
+			}
+		}
+
```
This is same as submarinerv1.HeadlessServicePod. We can remove the duplication

tpantelis replied:
yeah it differs by annotation key. I suggest to extract a method:

 `func (c *globalIngressIPController) addRulesForHeadlessService(ingressIP *submarinerv1.GlobalIngressIP, annotationKey string, ips []string) bool`

```
if addRulesForHeadlessService(ingressIP, headlessSvcEndpointsIP, ips) {
    return true
}
```

mkimuram replied:
Fixed in a bit different way (to be aligned to the other place). `targetType` is passed to add to different chains for pod ("SM-GN-EGRESS-HDLS-PODS") and endpoints ("SM-GN-EGRESS-HDLS-EPS") to keep existing behavior (and it helps  distinguishing the logs).

sridhargaddam reviewed (commented):
Thank you for the nice work @mkimuram.

Few things
1. Are you planning to push the Unit tests as part of the same PR?
2. We need changes to the RBAC to support listening for endpoints both in Submariner-operator and submariner-charts. I think you can go ahead and push these PRs first.
3. Did you test the end-to-end functionality and is it working fine?

tpantelis commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,153 @@
+	ingressIP.Spec = submarinerv1.GlobalIngressIPSpec{
+		Target:     submarinerv1.HeadlessServiceEndpoints,
+		ServiceRef: &corev1.LocalObjectReference{Name: c.svcName},
+		// TODO: Consider adding endpointsRef field to GlobalIngressIPSpec
```
I think we should definitely do this.

mkimuram replied:
Updated to add `endpointsRef`, however I started to think that it won't be necessary because name of Service and Endpoints are the same (I think that these fields are to track the full name for the case that it is truncated as a part of GlobalIngressIP's name, so ServiceRef field will be enough). 

If it is agreed, I will delete `endpointsRef` field.

tpantelis commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,153 @@
+		return ingressIP, false
+	}
+
+	// TODO: consider other conditions
```
I'm not clear on this comment.

mkimuram replied:
Just meant to remind considering other conditions to return here, like [this](https://github.com/submariner-io/submariner/blob/devel/pkg/globalnet/controllers/ingress_pod_controller.go#L118). Removed.

mkimuram reviewed (commented):
@sridhargaddam @tpantelis  Thank you for your review. I will reflect the comments and add unit tests.

mkimuram commented:
> We need changes to the RBAC to support listening for endpoints both in Submariner-operator and submariner-charts. I think you can go ahead and push these PRs first.

Created issues and PRs, and added descriptions on dependencies in [the first comment](https://github.com/submariner-io/submariner/pull/1558#issue-999555499) of this PR.

mkimuram commented:
@sridhargaddam @tpantelis 

Addressed review comments. PTAL

> Are you planning to push the Unit tests as part of the same PR?

Unit tests are added.

> Did you test the end-to-end functionality and is it working fine?

End-to-end functionality is tested manually as described in [here](https://github.com/submariner-io/submariner/pull/1558#issuecomment-921950088) and [here](https://github.com/submariner-io/submariner/pull/1558#issuecomment-922041105). I will add automated E2E tests later on top of https://github.com/submariner-io/submariner/pull/1555.

mkimuram commented:
Refactored and added E2E tests. The last three commits are for E2E tests, but only the tests in the last commit are for this PR's change. I will move the other E2E commits to another PR to make it clear.

Note that the tests in the last commits fails due to the lack of access right to endpointslice (and it seems that there is a bug in the cleanup logic of headless service without selector, because unrelated tests become fail after the tests). I will investigate and fix it.

sridhargaddam commented:
When this PR is ready for review, please remove the status from draft.

dfarrell07 commented:
I see this is marked as ready to review, thanks @mkimuram! It needs a manuel rebase.

mkimuram commented on pkg/globalnet/controllers/ingress_endpoints_controllers.go:
```diff
@@ -16,16 +16,18 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+//nolint:dupl // same logic to ingress_pod_controllers, but for a different class
```
Let me consider a bit more if I can fix it properly.

mkimuram reviewed (commented):
Rebased.

sridhargaddam commented on pkg/globalnet/constants/constants.go:
```diff
@@ -28,6 +28,7 @@ const (
 	// The following chains are added as part of GN 2.0 implementation.
 	SmGlobalnetEgressChainForPods            = "SM-GN-EGRESS-PODS"
 	SmGlobalnetEgressChainForHeadlessSvcPods = "SM-GN-EGRESS-HDLS-PODS"
+	SmGlobalnetEgressChainForHeadlessSvcEPs  = "SM-GN-EGRESS-HDLS-EPS"
```
Can we re-use the same chain SmGlobalnetEgressChainForHeadlessSvcPods?

mkimuram replied:
I think that we can. However, the name is a bit confusing and it makes rules for different purpose exist in one chain. If there is no side effect(like update issue), it would be clean to add a separate chain.

sridhargaddam replied:
I had a look at the code and thought about it for sometime and I think it should be okay even if we introduce a new chain in between. Can you please verify once in your local setup running this updated Globalnet image on an existing (v0.12.0) Globalnet setup. If you find that there is no issue, we can go with the new chain.

mkimuram replied:
@sridhargaddam Confirmed that it worked as expected. Please also see below details:

<details>

- Build the container image for this PR

```bash
git checkout issue/1537
make images
docker tag quay.io/submariner/submariner-globalnet:dev quay.io/submariner/submariner-globalnet:issue1537
```

- Deploy v0.12.0 submariner

```bash
git checkout v0.12.0

make deploy using=external-net,globalnet
```

- Confirm the nat chain for v0.12.0 (`SM-GN-EGRESS-HDLS-EPS` doesn't exist)

```bash
docker exec -it cluster1-worker iptables -t nat -S SUBMARINER-GN-EGRESS
# Warning: iptables-legacy tables present, use iptables-legacy to see them
-N SUBMARINER-GN-EGRESS
-A SUBMARINER-GN-EGRESS -j SUBMARINER-GN-MARK
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-PODS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-HDLS-PODS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-NS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-CLUSTER

docker exec -it cluster2-worker iptables -t nat -S SUBMARINER-GN-EGRESS
# Warning: iptables-legacy tables present, use iptables-legacy to see them
-N SUBMARINER-GN-EGRESS
-A SUBMARINER-GN-EGRESS -j SUBMARINER-GN-MARK
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-PODS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-HDLS-PODS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-NS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-CLUSTER
```

- Test that existing test pass
```bash
make e2e using=external-net,globalnet

• [SLOW TEST:46.771 seconds]
[external-dataplane] Connectivity
/go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivity
.go:41
  should be able to connect from an external app to a pod in a cluster
  /go/src/github.com/submariner-io/submariner/test/external/dataplane/connectivi
ty.go:44
------------------------------

JUnit path was configured: /go/src/github.com/submariner-io/submariner/output/e2
e-junit.xml

JUnit report was created: /go/src/github.com/submariner-io/submariner/output/e2e
-junit.xml

Ran 1 of 1 Specs in 46.822 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestE2E (46.82s)
PASS
```

- Update the globalnet image to the one built from this PR
(submariner-operator doesn't allow directly updating globalnet daemonset because it [reconciles the image](https://github.com/submariner-io/submariner-operator/blob/devel/controllers/submariner/globalnet_resources.go#L78). So, it seems that we need to update the `Submariner` CR).

```bash
kind load docker-image quay.io/submariner/submariner-globalnet:issue1537 --name cluster1
kind load docker-image quay.io/submariner/submariner-globalnet:issue1537 --name cluster2

export KUBECONFIG=output/kubeconfigs/kind-config-cluster1
kubectl get daemonset submariner-globalnet -n submariner-operator -o jsonpath='{range .spec.template.spec.containers[*]}{.name}{" "}{.image}{"\n"}{end}'
submariner-globalnet localhost:5000/submariner-globalnet:local

kubectl get submariner -n submariner-operator -o yaml > /tmp/submariner_cr.txt
kubectl edit submariner -n submariner-operator
kubectl get submariner -n submariner-operator -o yaml > /tmp/submariner_cr_after.txt

diff -urN  /tmp/submariner_cr.txt /tmp/submariner_cr_after.txt | grep -C 2 "imageOverrides"
     debug: false
     globalCIDR: 242.254.1.0/24
+    imageOverrides:
+      submariner-globalnet: quay.io/submariner/submariner-globalnet:issue1537
     namespace: submariner-operator

kubectl get daemonset submariner-globalnet -n submariner-operator -o jsonpath='{range .spec.template.spec.containers[*]}{.name}{" "}{.image}{"\n"}{end}'
submariner-globalnet quay.io/submariner/submariner-globalnet:issue1537
```

```bash
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
kubectl edit submariner -n submariner-operator

kubectl get daemonset submariner-globalnet -n submariner-operator -o jsonpath='{range .spec.template.spec.containers[*]}{.name}{" "}{.image}{"\n"}{end}'
submariner-globalnet quay.io/submariner/submariner-globalnet:issue1537
```

- Confirm the nat chain for this PR (`SM-GN-EGRESS-HDLS-EPS` exists in the right order)

```bash
docker exec -it cluster1-worker iptables -t nat -S SUBMARINER-GN-EGRESS
# Warning: iptables-legacy tables present, use iptables-legacy to see them
-N SUBMARINER-GN-EGRESS
-A SUBMARINER-GN-EGRESS -j SUBMARINER-GN-MARK
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-PODS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-HDLS-PODS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-HDLS-EPS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-NS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-CLUSTER

docker exec -it cluster2-worker iptables -t nat -S SUBMARINER-GN-EGRESS
# Warning: iptables-legacy tables present, use iptables-legacy to see them
-N SUBMARINER-GN-EGRESS
-A SUBMARINER-GN-EGRESS -j SUBMARINER-GN-MARK
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-PODS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-HDLS-PODS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-HDLS-EPS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-NS
-A SUBMARINER-GN-EGRESS -j SM-GN-EGRESS-CLUSTER
```

- Test that e2e test for this PR succeeds

```bash
git checkout issue/1537

go mod vendor

make "e2e" using="external-net globalnet"


Ran 15 of 28 Specs in 281.317 seconds
SUCCESS! -- 15 Passed | 0 Failed | 7 Pending | 6 Skipped
--- PASS: TestE2E (281.34s)
PASS

```

</details>

sridhargaddam commented on pkg/globalnet/controllers/gateway_monitor.go:
```diff
@@ -287,7 +288,12 @@ func (g *gatewayMonitor) startControllers() error {
 	}
 
-	c, err = NewServiceExportController(g.syncerConfig, podControllers)
+	endpointsControllers, err := NewIngressEndpointsControllers(g.syncerConfig)
```
nit: endpointsController and similarly NewIngressEndpointsController

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,149 @@
+	}
+
+	if c.ingressIPMap.Contains(ingressIP.Name) {
+		// Avoid assigning ingressIPs to endpoints that are not ready with an endpoint IP
```
I think this debug message is not relevant for endpoints

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,149 @@
+}
+
+func areEndpointsEqual(obj1, obj2 *unstructured.Unstructured) bool {
+	subsets1, _, _ := unstructured.NestedSlice(obj1.Object, "subsets")
```
Here we are checking the entire subsets whereas while creating the IngressIP, we only consider IPAddress from subsets[0] (Line 131). Should we check only subset[0]?

mkimuram replied:
In my understanding, it needs to be started here, because this controller is started via `IngressEndpointsControllers.start()` per Service basis. It is using the similar approach that ingress_pod_controllers.go does for ingress_pod_controller.go.

sridhargaddam commented on pkg/globalnet/controllers/service_export_controller_test.go:
```diff
@@ -253,6 +254,37 @@ func testHeadlessService() {
+		t.createServiceExport(t.createService(service))
+	})
+
+	When("an endpoints for an exported Service is created", func() {
```
nit: endpoints -> endpoint

mkimuram replied:
fixed.

mkimuram committed:
Implement globalnet for headless service without selector

Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

nyechiel commented:
@mkimuram can you look at the conflicts please?

mkimuram commented:
@nyechiel Thank you for taking a look at this PR. Actually, this PR has a dependency to other PR that is described in the first comment (and the depended PR has another depended PR that I'm working on). I will update this PR once they are merged.

mkimuram commented:
Created another branch https://github.com/mkimuram/submariner/commits/issue/1537v2 that is rebased on https://github.com/submariner-io/submariner/pull/1582#issuecomment-1043713012. I will update this PR based on it, once the PR is merged.

sridhargaddam commented:
@mkimuram PR that adds support in GN for services without selectors is now merged.

mkimuram commented:
I've rebased 1537v2 branch with the latest #1582 commits (E2E test PR) and will copy the branch once the E2E PR is merged.

mkimuram commented:
@nerdalert @anfredette

Please check https://github.com/submariner-io/submariner/pull/1558#discussion_r822977638 
Any feedback, like testing, review, and comments on use cases, is welcome.

mkimuram commented:
Rebased and resolved conflict.

mkimuram commented:
Let me check why the test failed and fix it.

mkimuram commented:
@dfarrell07 @sridhargaddam  I confirmed that accessing [via Service](https://github.com/submariner-io/submariner/blob/devel/test/external/dataplane/connectivity.go#L226) from external docker container [here](https://github.com/submariner-io/submariner/blob/devel/test/external/dataplane/connectivity.go#L233) is flaky in non-globalnet cases, even without this PR. The error happens because it takes some time before the service actually become ready to be accessed. Currently, there seems [no official way](https://github.com/kubernetes/kubernetes/issues/80828) to check if a service is actually ready. So, I've tried checking if there are any endpoints like [this workaround](https://github.com/openshift/cluster-logging-operator/pull/794/commits/d7c5aaa3cb0820908818605ad3c041b10808ce4d#diff-a4a97950ca409680f090d60ca871efdc3637f80567b3153d9d5af083b11e1cf4R158-R159) (see details below), but it still seems to fail with the change (maybe there is still a small amount of time between the endpoints field is updated and iptables rules are reflected). 

Therefore, I just added a sleep for this path in the last commit and it doesn't seem to flake now. Note that as for globalnet cases, they wait for globalIP to be assigned (and at that time they seem to be ready to be accessed), so they don't seem to flake.

<details>

```golang
diff --git a/test/external/dataplane/connectivity.go b/test/external/dataplane/connectivity.go
index 747b3ee..7a63750 100644
--- a/test/external/dataplane/connectivity.go
+++ b/test/external/dataplane/connectivity.go
@@ -19,6 +19,7 @@ limitations under the License.
 package dataplane
 
 import (
+	"context"
 	"fmt"
 	"sort"
 
@@ -26,6 +27,9 @@ import (
 	. "github.com/onsi/gomega"
 	"github.com/submariner-io/shipyard/test/e2e/framework"
 	"github.com/submariner-io/shipyard/test/e2e/tcp"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 const (
@@ -204,6 +208,7 @@ func testExternalConnectivity(p testParams) {
 		Command:       simpleHTTPServerCommand,
 	})
 	svc := np.CreateService()
+	awaitServiceReady(p.Cluster, svc.Namespace, svc.Name)
 
 	// Get handle for existing docker
 	docker := framework.New(extAppName)
@@ -298,3 +303,37 @@ func getNonGatewayClusterIndex(names []string) (framework.ClusterIndex, error) {
 	return framework.ClusterIndex(0),
 		fmt.Errorf("failed to find non-gateway-cluster (gateway-cluster name: %s, cluster names: %v)", clusterName, names)
 }
+
+func awaitServiceReady(cluster framework.ClusterIndex, namespace, svcName string) {
+	framework.AwaitUntil("await non-empty endpoints",
+		func() (interface{}, error) {
+			ep, err := framework.KubeClients[cluster].CoreV1().Endpoints(namespace).Get(context.TODO(), svcName, metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return nil, nil // nolint:nilnil // Returning nil value is intentional
+			}
+
+			return ep, err
+		},
+		func(result interface{}) (bool, string, error) {
+			if result == nil {
+				return false, fmt.Sprintf("endpoints %s/%s not found", namespace, svcName), nil
+			}
+
+			endpoints := result.(*v1.Endpoints)
+
+			var found bool
+			for _, subset := range endpoints.Subsets {
+				for _, addr := range subset.Addresses {
+					if addr.IP != "" {
+						found = true
+					}
+				}
+			}
+
+			if !found {
+				return false, fmt.Sprintf("no IP found in endpoints %s/%s", namespace, svcName), nil
+			}
+
+			return true, "", nil
+		})
+}
diff --git a/test/external/dataplane/gn_connectivity.go b/test/external/dataplane/gn_connectivity.go
index cea64e7..c9bfea4 100644
--- a/test/external/dataplane/gn_connectivity.go
+++ b/test/external/dataplane/gn_connectivity.go
@@ -332,6 +332,7 @@ func testGlobalNetExternalConnectivity(p testParams, g globalnetTestParams) {
 		Command:       simpleHTTPServerCommand,
 	})
 	svc := createSvc(p, np)
+	awaitServiceReady(p.Cluster, svc.Namespace, svc.Name)
 	p.Framework.CreateServiceExport(np.Config.Cluster, svc.Name)
 
 	// Get globalIPs for the network pod to use later
```

</details>

mkimuram commented:
Removed the workaround commit and rebased to include the fix for the flakey tests. Now it should be ready to be merged.

mkimuram commented:
@sridhargaddam @tpantelis Could you approve this PR?

sridhargaddam commented on pkg/globalnet/controllers/gateway_monitor.go:
```diff
@@ -390,6 +396,12 @@ func (g *gatewayMonitor) createGlobalnetChains() error {
 
+	klog.V(log.DEBUG).Infof("Install/ensure %s chain exists", constants.SmGlobalnetEgressChainForHeadlessSvcEPs)
+
+	if err := iptables.CreateChainIfNotExists(g.ipt, "nat", constants.SmGlobalnetEgressChainForHeadlessSvcEPs); err != nil {
```
Does this create issues during upgrades (i.e., what if the existing chains are already present and we are trying to insert a new chain in between)?

mkimuram replied:
I think that the jump rule to this chain is added by using `iptables.InsertUnique` with specifying the number to be inserted, so where this chain is inserted would be consistent regardless of whether the other chains already present.

However, sharing `SmGlobalnetEgressChainForHeadlessSvcPods` chain may be the other way to workaround it, as you mentioned.

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,142 @@
+
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:                "Ingress Endpoints syncer",
+		ResourceType:        &corev1.Endpoints{},
```
Can you please explain the impact on this code when the endpoints are deprecated/removed and replaced with EndpointSlices?

mkimuram replied:
If `Endpoints` are deprecated, we would need to re-implement this to use `EndpointSlices`. However, I'm not sure if there are fixed plan to deprecate it.

For this particular headless service without selector case, I think that we should use `Endpoints`, because `"The control plane automatically creates EndpointSlices for any Kubernetes Service that has a selector specified."` as described [here](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#endpointslice-resource). As I tested as below, `EndpointSlices` doesn't seem to be created automatically, for manually created `Endpoints`. 

<details>

```bash
kubectl get endpoints,endpointslices
NAME                   ENDPOINTS         AGE
endpoints/kubernetes   172.18.0.2:6443   2m39s

NAME                                        ADDRESSTYPE   PORTS   ENDPOINTS    AGE
endpointslice.discovery.k8s.io/kubernetes   IPv4          6443    172.18.0.2   2m39s

cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-vm
subsets:
  - addresses:
      - ip: 192.168.4.2
    ports:
      - port: 80
EOF

kubectl get endpoints,endpointslices
NAME                   ENDPOINTS         AGE
endpoints/kubernetes   172.18.0.2:6443   3m48s
endpoints/test-vm      192.168.4.2:80    4s

NAME                                        ADDRESSTYPE   PORTS   ENDPOINTS    AGE
endpointslice.discovery.k8s.io/kubernetes   IPv4          6443    172.18.0.2   3m48s
```
</details>

sridhargaddam replied:
Ok, thanks @mkimuram

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,142 @@
+	})
+
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating the syncer")
```
nit: "error creating the endpoints syncer"

mkimuram replied:
Fixed as suggested.

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,142 @@
+	}
+
+	if err := controller.Start(); err != nil {
+		return nil, errors.Wrap(err, "error creating the syncer")
```
"error starting the endpoint syncer"

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,142 @@
+		return nil, errors.Wrap(err, "error creating the syncer")
+	}
+
+	if err := controller.Start(); err != nil {
```
Any specific reason for starting the controller here instead of in gateway_monitor.go Line 316?

mkimuram replied:
In my understanding, it needs to be started here, because this controller is started via `IngressEndpointsControllers.start()` per Service basis. It is using the similar approach that ingress_pod_controllers.go does for ingress_pod_controller.go.

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,203 @@
+		annotations := ingressIP.GetAnnotations()
+		if epIP, ok := annotations[headlessSvcEndpointsIP]; ok {
+			if _, used := epIPs[epIP]; used {
+				// ingressIP is still used
```
If the ingressIP is still used, should we remove it from the epIP list? As otherwise we seem to create new IngressIPs in Line 152. Am I missing something?

mkimuram replied:
I thought that it is safe to ensure the ingressIP resource to exist, but I noticed that calling `create` without checking already exist error doesn't do the intended works. So, it would be better removing it from epIP as you mentioned. Thank you for pointing it out.

sridhargaddam commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,203 @@
+func newIngressIP(name, namespace, epIP string) *submarinerv1.GlobalIngressIP {
+	return &submarinerv1.GlobalIngressIP{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("ep-%.44s-%.15s", name, epIP),
```
So if the user creates two different endpoints for different services where the initial 44 chars of names match and epIP matches, we would have a problem here. Is this correct? Its a very corner case, but just want to hear from you if I'm reading the code correctly.

mkimuram replied:
Correct. It's similar restriction to [the one for pod](https://github.com/submariner-io/submariner/blob/devel/pkg/globalnet/controllers/ingress_pod_controller.go#L108), but much shorter than than the pod one. We might be able to use hash, but it would make it harder to recognize what endpoints this ingress is for. WDYT?

sridhargaddam replied:
Agree, using a hast makes it difficult to recognise it. Since its a very corner case, lets continue this way.

sridhargaddam commented on pkg/globalnet/controllers/iptables/iface.go:
```diff
@@ -105,13 +112,13 @@ func (i *ipTables) ipTableChainExists(table, chain string) (bool, error) {
-		return fmt.Errorf("globalIP %q or podIP %q cannot be empty", globalIP, podIP)
+func (i *ipTables) AddIngressRulesForHeadlessSvc(globalIP, ip string, targetType TargetType) error {
+	if globalIP == "" || ip == "" {
+		return fmt.Errorf("globalIP %q or %s IP %q cannot be empty", targetType, globalIP, ip)
```
```suggestion
		return fmt.Errorf("globalIP %q or %s IP %q cannot be empty", globalIP, targetType, ip)
```

mkimuram replied:
Fixed as suggested.

sridhargaddam commented on pkg/globalnet/controllers/iptables/iface.go:
```diff
@@ -120,14 +127,14 @@ func (i *ipTables) AddIngressRulesForHeadlessSvcPod(globalIP, podIP string) erro
-		return fmt.Errorf("globalIP %q or podIP %q cannot be empty", globalIP, podIP)
+func (i *ipTables) RemoveIngressRulesForHeadlessSvc(globalIP, ip string, targetType TargetType) error {
+	if globalIP == "" || ip == "" {
+		return fmt.Errorf("globalIP %q or %s IP %q cannot be empty", targetType, globalIP, ip)
```
```suggestion
		return fmt.Errorf("globalIP %q or %s IP %q cannot be empty", globalIP, targetType, ip)
```

mkimuram replied:
Fixed as suggested.

sridhargaddam reviewed (changes requested):
Thanks for the PR @mkimuram.

mkimuram reviewed (commented):
@sridhargaddam Thank you for your review and comments. I put some comments back and will update the code latter.

mkimuram commented on pkg/globalnet/controllers/uninstall.go:
```diff
@@ -52,6 +52,7 @@ func UninstallDataPath() {
 		// The chains have to be deleted in a specific order.
 		constants.SmGlobalnetEgressChainForCluster,
 		constants.SmGlobalnetEgressChainForHeadlessSvcPods,
+		constants.SmGlobalnetEgressChainForHeadlessSvcEPs,
```
Also, added uninstall code here. 

By this change,  this version of subctl comand starts to try deleting this chain, while old globalnet deployment doesn't have this chain. However,  uninstallation continues even `FlushIPTableChain` fails, so there won't be issue by version mismatch. 

Actually, [`ClearChain`](https://github.com/coreos/go-iptables/blob/master/iptables/iptables.go#L427) called from `FlushIPTableChain` doesn't seem to return error even when the chain doesn't exist already (it creates new one and flush). So, it would be safe to just add the chain to the list to be deleted.

sridhargaddam reviewed (approved):
Thanks for addressing the review comments @mkimuram.

@tpantelis @anfredette can you please review.

sridhargaddam commented:
@mkimuram The last commit can be squashed with the previous one as there is a linting error.

tpantelis commented on pkg/globalnet/controllers/base_controllers.go:
```diff
@@ -73,11 +73,21 @@ func (c *baseSyncerController) Start() error {
+
 	c.resourceSyncer.Reconcile(func() []runtime.Object {
-		objList, err := client.List(context.TODO(), metav1.ListOptions{LabelSelector: labelSelector})
+		objList, err := client.List(context.TODO(), listOptions)
```
We can simplify this.

```suggestion
		objList, err := client.List(context.TODO(), metav1.ListOptions{
		    LabelSelector: labelSelector,
		    FieldSelector: fieldSelector,
		})
```

tpantelis commented on pkg/globalnet/controllers/global_ingressip_controller.go:
```diff
@@ -214,24 +224,42 @@ func (c *globalIngressIPController) onCreate(ingressIP *submarinerv1.GlobalIngre
+				return true
+			}
+
+			tType = iptables.EndpointsTarget
```
We can simplify to:

```Go
var annotationKey string
var tType iptables.TargetType

if ingressIP.Spec.Target == submarinerv1.HeadlessServicePod {
    annotationKey = headlessSvcPodIP
    tType = iptables.PodTarget
} else if ingressIP.Spec.Target == submarinerv1.HeadlessServiceEndpoints {
    annotationKey = headlessSvcEndpointsIP
    tType = iptables.EndpointsTarget
}

target = ingressIP.GetAnnotations()[annotationKey]
if target == "" {
    _ = c.pool.Release(ips...)

    klog.Warningf("%q annotation is missing on %q", annotationKey, key)

    return true
}
```

tpantelis commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,206 @@
+	klog.Infof("%q ingress Endpoints %s for service %s", op, key, c.svcName)
+
+	// Get list of current endpoint IPs
+	epIPs := map[string]bool{}
```
```suggestion
	usedEpIPs := map[string]bool{}
```

tpantelis commented on pkg/globalnet/controllers/ingress_endpoints_controller.go:
```diff
@@ -0,0 +1,206 @@
+		annotations := ingressIP.GetAnnotations()
+		if epIP, ok := annotations[headlessSvcEndpointsIP]; ok {
+			// ingressIP is still used
+			if _, used := epIPs[epIP]; used {
```
```suggestion
			if usedEpIPs[epIP] {
```

tpantelis reviewed (changes requested):
Looks good - just a couple minor things.

mkimuram committed:
Simplify codes and improve readability

Signed-off-by: Masaki Kimura <masaki.kimura@hitachivantara.com>

mkimuram commented:
@tpantelis Thank you for your review. Fixed as suggested.

Note that `"Linting / Commit Message(s)"` and `"Linting / Markdown Links (modified files)"` are constantly failing, so "not required". Also, `"Upgrade / Latest Release to Latest Version"` seems flaking (but succeeded this time).
